vpr/src/analytical_place/full_legalizer.cpp

```diff
@@ -0,0 +1,406 @@
+                    ++it;
+                }
+            }
+            // Once all molecules have been inserted, clean the cluster.
+            cluster_legalizer.clean_cluster(new_cluster_id);
+        }
+    }
+
+    // Check and output the clustering.
+    std::unordered_set<AtomNetId> is_clock = alloc_and_load_is_clock();
+    check_and_output_clustering(cluster_legalizer, packer_opts_, is_clock, arch_);
+    // Reset the cluster legalizer. This is required to load the packing.
+    cluster_legalizer.reset();
+    // Regenerate the clustered netlist from the file generated previously.
+    // FIXME: This writing and loading from a file is wasteful. Should generate
+    //        the clusters directly from the cluster legalizer.
+    vpr_load_packing(vpr_setup_, *arch_);
+    load_cluster_constraints();
+    const ClusteredNetlist& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+
+    // Verify the packing and print some info
+    check_netlist(packer_opts_.pack_verbosity);
+    writeClusteredNetlistStats(vpr_setup_.FileNameOpts.write_block_usage);
+    print_pb_type_count(clb_nlist);
+
+    // PLACING:
+    // Create a lookup from the AtomBlockId to the APBlockId
+    vtr::vector<AtomBlockId, APBlockId> atom_to_ap_block(atom_netlist_.blocks().size());
+    for (APBlockId ap_blk_id : ap_netlist_.blocks()) {
+        const t_pack_molecule* blk_mol = ap_netlist_.block_molecule(ap_blk_id);
+        for (AtomBlockId atom_blk_id : blk_mol->atom_block_ids) {
+            // Ensure that this block is not in any other AP block. That would
+            // be weird.
+            VTR_ASSERT(!atom_to_ap_block[atom_blk_id].is_valid());
+            atom_to_ap_block[atom_blk_id] = ap_blk_id;
+        }
+    }
+    // Move the clusters to where they want to be first.
+    // TODO: The fixed clusters should probably be moved first for legality
+    //       reasons.
+    APClusterPlacer ap_cluster_placer;
+    std::vector<ClusterBlockId> unplaced_clusters;
+    for (ClusterBlockId cluster_blk_id : clb_nlist.blocks()) {
+        // Assume that the cluster will always want to be placed wherever the
+        // first atom in the cluster wants to be placed.
+        // FIXME: This assumption does not always hold! Will need to unify the
+        //        cluster legalizer and the clustered netlist!
+        const std::unordered_set<AtomBlockId>& atoms_in_cluster = g_vpr_ctx.clustering().atoms_lookup[cluster_blk_id];
+        VTR_ASSERT(atoms_in_cluster.size() > 0);
+        AtomBlockId first_atom_blk = *atoms_in_cluster.begin();
+        APBlockId first_ap_blk = atom_to_ap_block[first_atom_blk];
+        size_t blk_x_loc = p_placement.block_x_locs[first_ap_blk];
+        size_t blk_y_loc = p_placement.block_y_locs[first_ap_blk];
+        size_t blk_layer_num = p_placement.block_layer_nums[first_ap_blk];
+        size_t blk_sub_tile = p_placement.block_sub_tiles[first_ap_blk];
+        t_physical_tile_loc tile_loc(blk_x_loc, blk_y_loc,blk_layer_num);
```

Comment:
I think you should have a helper function that takes a double loc (ap_loc? pl_cont_loc?) and snaps it to the nearest grid tile (taking into account anchor points).  This code will round down, and also may try to place things at locations that aren't the anchor point of a big block.  A helper function would encapsulate the extra checks etc.